Modules/ReverseProxy/Websocket/RawWebsocketConnection.cs

```diff
@@ -0,0 +1,175 @@
+                continue;
+
+            upgradeRequest += $"{header.Key}: {header.Value}\r\n";
+        }
+
+        upgradeRequest += "\r\n";
+
+        Console.WriteLine($"Upgrade Request (Server -> Upstream): {upgradeRequest}");
+
+        // Writes and flushes
+        await Pipe.Output.WriteAsync(Encoding.UTF8.GetBytes(upgradeRequest), token);
+
+        // Read the handshake response until \r\n\r\n
+        while (true)
+        {
+            var result = await Pipe.Input.ReadAsync(token);
+            var buffer = result.Buffer;
+
+            if (result.IsCompleted && buffer.Length == 0)
+            {
+                throw new InvalidOperationException(
+                    "Connection closed before full WebSocket handshake was received.");
+            }
+
+            var sequenceReader = new SequenceReader<byte>(buffer);
+
+            SequencePosition consumed = buffer.Start;
+            SequencePosition examined = buffer.End;
+
+            try
+            {
+                if (sequenceReader.TryReadTo(
+                        out ReadOnlySequence<byte> headersSequence,
+                        "\r\n\r\n"u8,
+                        advancePastDelimiter: true))
+                {
+                    consumed = sequenceReader.Position; // consumed up to here
+                    examined = consumed;
+
+                    // Validate status line + headers
+                    var headersText = Encoding.ASCII.GetString(headersSequence.ToArray());
+
+                    Console.WriteLine($"Received Handshake (Upstream -> Server): {headersText}");
+
+                    return headersText.StartsWith("HTTP/1.1 101", StringComparison.Ordinal);
+                }
+                else
+                {
+                    if (result.IsCompleted)
+                    {
+                        throw new InvalidOperationException("Connection closed before full WebSocket handshake was received.");
+                    }
+                }
+            }
+            finally
+            {
+                // TODO: Does it make sense to advance in case of exception?
+                Pipe.Input.AdvanceTo(consumed, examined);
+            }
+        }
+    }
+
+    private static (string host, int port, bool secure, string route) GetHostPortAndSecurity(string url)
+    {
+        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri))
+            throw new ArgumentException($"Invalid URL: {url}", nameof(url));
+
+        bool secure = uri.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase) ||
+            uri.Scheme.Equals("wss", StringComparison.OrdinalIgnoreCase);
+
+        int port = !uri.IsDefaultPort
+            ? uri.Port
+            : uri.Scheme.ToLowerInvariant() switch
+            {
+                "https" => 443,
+                "wss" => 443,
+                "http" => 80,
+                "ws" => 80,
+                _ => 0
+            };
+
+        // Extract only the path, ensure at least "/"
```

Comment:
Why no query parameters? Currently they will not be relayed - I think they should.